pyccel/parser/syntactic.py

```diff
@@ -439,65 +452,65 @@ def _visit_UnitaryOperatorNode(self, stmt):
+
         if stmt.value == '+':
-            return Add(first, second)
+            expr = PyccelAdd(first, second)
+            return change_priority(expr)
+
         elif stmt.value == '*':
+            expr = PyccelMul(first, second)
+            return change_priority(expr)
 
-            if isinstance(first, (PythonTuple, Tuple, List)):
-                return Dlist(first, second)
-            return Mul(first, second)
         elif stmt.value == '-':
+            expr = PyccelMinus(first, second)
+            return change_priority(expr)
 
-            if isinstance(stmt.second, BinaryOperatorNode) \
-                and isinstance(second, (Add, Mul)):
-                args = second.args
-                second = second._new_rawargs(-args[0], args[1])
-            else:
-                second = Mul(-1, second)
-            return Add(first, second)
         elif stmt.value == '/':
-            if isinstance(second, Mul) and isinstance(stmt.second,
-                                           BinaryOperatorNode):
-                args = list(second.args)
-                second = Pow(args[0], Integer(-1))
-                second = Mul(second, args[1])
-            else:
-                second = Pow(second, Integer(-1))
-            return Mul(first, second)
+            expr = PyccelDiv(first, second)
+            return change_priority(expr)
 
         elif stmt.value == '**':
+            expr = PyccelPow(first, second)
+            return change_priority(expr)
 
-            return Pow(first, second)
         elif stmt.value == '//':
-
-            if isinstance(second, Mul) and isinstance(stmt.second,
-                                           BinaryOperatorNode):
-                args = second.args
-                second = Pow(args[0], Integer(-1))
-                first =  floor(Mul(first, second), evaluate=False)
-                return Mul(first, args[1])
-            else:
-                second = Pow(second, Integer(-1))
-                return floor(Mul(first, second), evaluate=False)
+            expr = PyccelFloorDiv(first, second)
+            return change_priority(expr)
 
         elif stmt.value == '%':
-            return Mod(first, second)
+            expr = PyccelMod(first, second)
+            return change_priority(expr)
         else:
             msg = 'unknown/unavailable BinaryOperatorNode {node}'
             msg = msg.format(node=type(stmt.value))
             raise PyccelSyntaxError(msg)
 
-
     def _visit_BooleanOperatorNode(self, stmt):
 
         first = self._visit(stmt.first)
         second = self._visit(stmt.second)
+
         if stmt.value == 'and':
-            return And(first, second, evaluate=False)
+            if isinstance(second, PyccelOr):
+                args  = second.args
+                if isinstance(second, PyccelAnd):
+                    return PyccelAnd(first, *args)
+                first = PyccelAnd(first, args[0]  )
+                return  PyccelOr (first, *args[1:])
```

Comment:
This seems to carry out this conversion
And(a,Or(b,c)) -> Or(And(a,b), c)
These are not equivalent

Reply:
this carry out this conversion `a and b or c`

Reply:
So it changes the reading order to the correct left to right?
Can PyccelAnd and PyccelOr classes have more than 2 arguments?
If so then does the logic still work for longer expressions?

Reply:
not for the moment but maybe in the futur

Reply:
I am also puzzled by this function. Maybe @saidctb wanted to implement this identity instead?
`a and (b or c) == (a and b) or (a and c)`

Reply:
here there is no brackets, it only changes the evaluation from left to right instead of right to left
so when we have something like `a or b and c` redbaron gives us `Or(a,And(b,c))` and i change that to `And(Or(a,b),c)` the printing of these two expression is the same because there is no brackets here

Reply:
Cool. Thanks